src/codegen/yul/builtin.rs

```diff
@@ -0,0 +1,514 @@
+            Expression::Builtin(*loc, vec![Type::Value], Builtin::Balance, vec![addr])
+        }
+
+        YulBuiltInFunction::Caller => {
+            Expression::Builtin(*loc, vec![Type::Address(true)], Builtin::Sender, vec![])
+        }
+
+        YulBuiltInFunction::CallValue => {
+            Expression::Builtin(*loc, vec![Type::Value], Builtin::Value, vec![])
+        }
+
+        YulBuiltInFunction::SelfDestruct => {
+            let recipient = expression(&args[0], contract_no, ns, vartab, cfg, opt).cast(&Type::Address(true), ns);
+            cfg.add(vartab, Instr::SelfDestruct { recipient });
+            Expression::Poison
+        }
+
+        YulBuiltInFunction::Invalid => {
+            cfg.add(vartab, Instr::AssertFailure { expr: None });
+            Expression::Poison
+        }
+
+        YulBuiltInFunction::GasPrice => {
+            Expression::Builtin(*loc, vec![Type::Uint(64)], Builtin::Gasprice, vec![])
+        }
+
+        YulBuiltInFunction::BlockHash => {
+            let arg = expression(&args[0], contract_no, ns, vartab, cfg, opt).cast(&Type::Uint(64), ns);
+            Expression::Builtin(*loc, vec![Type::Uint(256)], Builtin::BlockHash, vec![arg])
+        }
+
+        YulBuiltInFunction::CoinBase => {
+            Expression::Builtin(*loc, vec![Type::Address(false)], Builtin::BlockCoinbase, vec![])
+        }
+
+        YulBuiltInFunction::Timestamp => {
+            Expression::Builtin(*loc, vec![Type::Uint(64)], Builtin::Timestamp, vec![])
+        }
+
+        YulBuiltInFunction::Number => {
+            Expression::Builtin(*loc, vec![Type::Uint(64)], Builtin::BlockNumber, vec![])
+        }
+
+        YulBuiltInFunction::Difficulty => {
+            Expression::Builtin(*loc, vec![Type::Uint(256)], Builtin::BlockDifficulty, vec![])
+        }
+
+        YulBuiltInFunction::GasLimit => {
+            Expression::Builtin(*loc, vec![Type::Uint(64)], Builtin::GasLimit, vec![])
+        }
+    }
+}
+
+/// Process arithmetic operations with two arguments
+fn process_binary_arithmetic(
+    loc: &pt::Loc,
+    builtin_ty: &YulBuiltInFunction,
+    args: &[ast::YulExpression],
+    contract_no: usize,
+    ns: &Namespace,
+    vartab: &mut Vartable,
+    cfg: &mut ControlFlowGraph,
+    opt: &Options,
+) -> Expression {
+    let left = expression(&args[0], contract_no, ns, vartab, cfg, opt);
+    let right = expression(&args[1], contract_no, ns, vartab, cfg, opt);
+    let (left, right) = equalize_types(left, right, ns);
+
+    match builtin_ty {
+        YulBuiltInFunction::Add => {
+            Expression::Add(*loc, left.ty(), true, Box::new(left), Box::new(right))
+        }
+        YulBuiltInFunction::Sub => {
+            Expression::Subtract(*loc, left.ty(), true, Box::new(left), Box::new(right))
+        }
+        YulBuiltInFunction::Mul => {
+            Expression::Multiply(*loc, left.ty(), true, Box::new(left), Box::new(right))
```

Comment:
The overflow checking is always off. I thought that `true` would indicate that because of this line:

https://github.com/hyperledger-labs/solang/blob/7f113eba256dbd38948c9d4cd41db8e034621758/src/emit/mod.rs#L5313

If unchecked is true, then we build an operation with overflow detection.

Reply:
Yes, you are right. Somehow I thought that the `bool` field meant "checked" and not "unchecked"